nativelib/src/main/resources/gc/immix/Allocator.c

```diff
@@ -0,0 +1,259 @@
+ * @return
+ */
+Allocator *Allocator_create(word_t *heapStart, int blockCount) {
+    Allocator *allocator = malloc(sizeof(Allocator));
+    allocator->heapStart = heapStart;
+
+    BlockList_init(&allocator->recycledBlocks, heapStart);
+    BlockList_init(&allocator->freeBlocks, heapStart);
+
+    // Init the free block list
+    allocator->freeBlocks.first = (BlockHeader *)heapStart;
+    BlockHeader *lastBlockHeader =
+        (BlockHeader *)(heapStart + ((blockCount - 1) * WORDS_IN_BLOCK));
+    allocator->freeBlocks.last = lastBlockHeader;
+    lastBlockHeader->header.nextBlock = LAST_BLOCK;
+
+    // Block stats
+    allocator->blockCount = (uint64_t)blockCount;
+    allocator->freeBlockCount = (uint64_t)blockCount;
+    allocator->recycledBlockCount = 0;
+
+    Allocator_initCursors(allocator);
+
+    return allocator;
+}
+
+/**
+ * The Allocator needs one free block for overflow allocation and a free or
+ * recyclable block for normal allocation.
+ *
+ * @param allocator
+ * @return `true` if there are enough block to initialise the cursors, `false`
+ * otherwise.
+ */
+bool Allocator_canInitCursors(Allocator *allocator) {
+    return allocator->freeBlockCount >= 2 ||
+           (allocator->freeBlockCount == 1 &&
+            allocator->recycledBlockCount > 0);
+}
+
+/**
+ *
+ * Used to initialise the cursors of the Allocator after collection
+ *
+ * @param allocator
+ */
+void Allocator_initCursors(Allocator *allocator) {
+
+    // Init cursor
+    allocator->block = NULL;
+    allocator->cursor = NULL;
+    allocator->limit = NULL;
+
+    getNextLine(allocator);
+
+    // Init large cursor
+    assert(!BlockList_isEmpty(&allocator->freeBlocks));
+
+    BlockHeader *largeHeader =
+        BlockList_removeFirstBlock(&allocator->freeBlocks);
+    allocator->largeBlock = largeHeader;
+    allocator->largeCursor = Block_getFirstWord(largeHeader);
+    allocator->largeLimit = Block_getBlockEnd(largeHeader);
+}
+
+/**
+ * Heuristic that tells if the heap should be grown or not.
+ */
+bool Allocator_shouldGrow(Allocator *allocator) {
+    uint64_t unavailableBlockCount =
+        allocator->blockCount -
+        (allocator->freeBlockCount + allocator->recycledBlockCount);
+
+#ifdef DEBUG_PRINT
+    printf("\n\nBlock count: %llu\n", allocator->blockCount);
+    printf("Unavailable: %llu\n", unavailableBlockCount);
+    printf("Free: %llu\n", allocator->freeBlockCount);
+    printf("Recycled: %llu\n", allocator->recycledBlockCount);
+    fflush(stdout);
+#endif
+
+    return allocator->freeBlockCount < allocator->blockCount / 3 ||
+           4 * unavailableBlockCount > allocator->blockCount ||
+           allocator->freeMemoryAfterCollection * 2 <
+               allocator->blockCount * BLOCK_TOTAL_SIZE;
+}
+
+/**
+ * Overflow allocation uses only free blocks, it is used when the bump limit of
+ * the fast allocator is too small to fit
+ * the block to alloc.
+ */
+word_t *overflowAllocation(Allocator *allocator, size_t size) {
```

Comment:
Do you want to add the module name to all functions ? I did it only for "public" (defined in .h) functions.